libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/MapboxTripStarter.kt

```diff
@@ -0,0 +1,171 @@
+
+        // Initialize the options to be aware of the location permissions
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        val granted = PermissionsManager.areLocationPermissionsGranted(context)
+        optionsFlow.update { it.toBuilder().isLocationPermissionGranted(granted).build() }
+
+        // Observe any changes to the options
+        coroutineScope.launch {
+            optionsFlow.collect { options ->
+                onStarterOptionsChanged(mapboxNavigation, options)
+            }
+        }
+    }
+
+    /**
+     * Signals that the [mapboxNavigation] instance is being detached.
+     * @param mapboxNavigation
+     */
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope.cancel()
+        onTripDisabled(mapboxNavigation)
+        this.mapboxNavigation = null
+    }
+
+    /**
+     * Get the options that are currently set.
+     */
+    fun getOptions(): MapboxTripStarterOptions = optionsFlow.value
+
+    /**
+     * Set new options.
+     */
+    fun setOptions(options: MapboxTripStarterOptions) = apply {
+        checkOptions(options)
+        this.optionsFlow.value = options
+    }
+
+    /**
+     * Apply changes to existing options.
+     */
+    fun update(function: (MapboxTripStarterOptions.Builder) -> MapboxTripStarterOptions.Builder) {
+        this.optionsFlow.update {
+            val nextOptions = function.invoke(it.toBuilder()).build()
+            checkOptions(nextOptions)
+            nextOptions
+        }
+    }
+
+    /**
+     * Throws an error if location permissions are set to true but the location permissions are
+     * not actually granted.
+     */
+    private fun checkOptions(options: MapboxTripStarterOptions) {
+        val mapboxNavigation = this.mapboxNavigation
+        checkNotNull(mapboxNavigation) {
+            "MapboxTripStarter cannot be used while MapboxNavigation is detached."
+        }
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        val granted = options.isLocationPermissionGranted
+        if (granted && !PermissionsManager.areLocationPermissionsGranted(context)) {
+            error(
+                "updateLocationPermissions can only be set to true when location permissions" +
+                    " have been accepted."
+            )
+        }
+    }
+
+    private fun onStarterOptionsChanged(
+        mapboxNavigation: MapboxNavigation,
+        options: MapboxTripStarterOptions
+    ) {
+        if (options.tripType == MAPBOX_TRIP_STARTER_REPLAY_ROUTE) {
+            onReplayTripEnabled(mapboxNavigation)
+        } else if (options.tripType == MAPBOX_TRIP_STARTER_FOLLOW_DEVICE &&
+            options.isLocationPermissionGranted
+        ) {
+            onTripSessionEnabled(mapboxNavigation)
+        } else {
+            onTripDisabled(mapboxNavigation)
+        }
+    }
+
+    private fun onReplayTripEnabled(mapboxNavigation: MapboxNavigation) {
```

Comment:
> IMO onReplayTripEnabled implies that replay has already been enabled

I'm not sure what you mean yet because this private function is called when replay has already been enabled. The functions are responding to the `options.tripType`

``` kotlin
if (options.tripType == MAPBOX_TRIP_STARTER_REPLAY_ROUTE) {
            onReplayTripEnabled(mapboxNavigation)
```

Reply:
I added this function just to make the `onStarterOptionsChanged` more clear 😢 . 

i don't think `enableReplayTrip` is more clear, because it is already enabled at this point

Reply:
Added the `enableReplayTrip` function cause I think, not having it was causing confusion